custom-bash-prompt/custom-bash-prompt.c

```diff
@@ -0,0 +1,192 @@
+ *
+ * @return Exit code to use for this program.
+ *****************************************************************************/
+int report_command_status(char const *last_command, int exit_code, long long begin)
+{
+    long long end = get_timestamp();
+    long long delay = end - begin;
+    int this_exit_code = delay > 10000 ? EXIT_SUCCESS : EXIT_FAILURE;
+    LOG("Command '%s' exited with code %d in %lld ms.", last_command, exit_code, delay);
+    if (delay <= 5000)
+    {
+        return this_exit_code;
+    }
+
+    // Remove the initial part (index and timestamp) of the command. Then
+    // allocate enough space to write it and some additional information.
+    last_command = strstr(last_command, RIGHT_SQUARE_BRACKET) + 2;
+    char *report = malloc((strlen(last_command) + 32) * sizeof *report);
+    char *report_ptr = report;
+
+    report_ptr += sprintf(report_ptr, "%s ", last_command);
+    if (exit_code == 0)
+    {
+        report_ptr += sprintf(report_ptr, bgreen "✓" rst " ");
+    }
+    else
+    {
+        report_ptr += sprintf(report_ptr, bred "✗" rst " ");
+    }
+    int milliseconds = delay % 1000;
+    int seconds = (delay /= 1000) % 60;
+    int minutes = (delay /= 60) % 60;
+    int hours = delay / 60;
+    LOG("Calculated delay is %d h %d m %d s %d ms.", hours, minutes, seconds, milliseconds);
+    if (hours > 0)
+    {
+        report_ptr += sprintf(report_ptr, "%d h ", hours);
+    }
+    if (hours > 0 || minutes > 0)
+    {
+        report_ptr += sprintf(report_ptr, "%d m ", minutes);
+    }
+    sprintf(report_ptr, "%d.%03d s", seconds, milliseconds);
+
+    // Ensure that the text is right-aligned even if it spans multiple lines.
+    // Since there are non-printable characters in the string, compensate for
+    // the width.
+    int report_len = (int)strlen(report);
+    int columns = atoi(getenv("COLUMNS"));
+    columns = columns - report_len % columns + report_len + 14;
+    LOG("Padding report of length %d to %d columns (adjusted).", report_len, columns);
+    fprintf(stderr, "%*s\n", columns, report);
+
+    return this_exit_code;
+}
+
+/******************************************************************************
+ * Update the title of the current terminal window. This should also
+ * automatically update the title of the current terminal tab.
+ *
+ * @param pwd Current working directory.
+ *****************************************************************************/
+void update_terminal_title(char const *pwd)
+{
+    char const *short_pwd = strrchr(pwd, '/') + 1;
+    fprintf(stderr, ESCAPE RIGHT_SQUARE_BRACKET "0;%s/" BELL, short_pwd);
+}
+
+/******************************************************************************
+ * Obtain information about the status of the current Git repository.
+ *****************************************************************************/
+char const *get_git_info(void)
+{
+    return dgreen "placeholder" rst;
+}
```

Comment:
Consider setting this to `$(__git_ps1 "   %s")`, so that the command is evaluated when `PS1` is set.